### PR TITLE
SK cuda

### PR DESCRIPTION
**Pull request by sakthipriyas**

### CODE REVIEW

File: psrdada_cpp/effelsberg/edd/SKRfiReplacement.hpp [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+class SKRfiReplacement{
+public:
+    /**
+     * @brief    constructor
+     *
+     * @param(in)     rfi_status    rfi_status of input data
+     *
+     */
+    SKRfiReplacement(const std::vector<int> &rfi_status);
```
Comment (TobiasWinchen) [inline psrdada_cpp/effelsberg/edd/SKRfiReplacement.hpp:27]: The RFI Status should not be passed on construction as it is specific to one data block. Pass it for very call of replac_rfi_data

File: psrdada_cpp/effelsberg/edd/SKRfiReplacementCuda.cuh [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+#include <thrust/execution_policy.h>
+#include <thrust/random/normal_distribution.h>
+#include <thrust/random/linear_congruential_engine.h>
+#include <nvToolsExt.h>
+
+namespace psrdada_cpp {
+namespace effelsberg {
+namespace edd {
+
+#define DEFAULT_NUM_CLEAN_WINDOWS 1 //number of clean windows used for computing DataStatistics
```
Comment (TobiasWinchen) [inline psrdada_cpp/effelsberg/edd/SKRfiReplacementCuda.cuh:18]: Why is this a global constant and not a parameter?

File: psrdada_cpp/effelsberg/edd/SpectralKurtosis.hpp [hunk 3]
```diff
@@ -0,0 +1,58 @@
+#include "psrdada_cpp/common.hpp"
+#include <complex>
+#include <vector>
+#include <numeric>
+
+namespace psrdada_cpp {
+namespace effelsberg {
+namespace edd {
+
+#define DEFAULT_SK_MIN 0.9
```
Comment (TobiasWinchen) [inline psrdada_cpp/effelsberg/edd/SpectralKurtosis.hpp:10]: Define in constructor only - what is the value of having additional macros?

**Review by TobiasWinchen — CHANGES_REQUESTED** [review]
In addition to few comments on specific cod lines, some general things that should be addressed before merging:
- Include guards are missing
- If the thrust version is obsolete, it should be removed and the naming adopted. The same goes for the CPU version. If it is only needed for testing, it should be moved to the test folder.
- The ReadData method is not used. It should be explained why it is needed or be removed. 
- All GPU compute is currently done in the default stream. The user of the class needs to b able to set the stream stream.

In the SKReplacement class several internal parameter-less methods are used (get_clean_window_indices(),  get_rfi_window_indices, compute_clean_data_statistics) that modify the internal state of the class. This makes the code hard to follow.

File: psrdada_cpp/effelsberg/edd/src/SKRfiReplacementCuda.cu [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+{
+    BOOST_LOG_TRIVIAL(info) << "Destroying SKRfiReplacementCuda instance..\n";
+}
+
+void SKRfiReplacementCuda::init()
+{
+    BOOST_LOG_TRIVIAL(info) << "initializing data_members of SKRfiReplacementCuda class..\n";
+    _nwindows = _rfi_status.size();
+    _rfi_window_indices.reserve(_nwindows);
+    get_rfi_window_indices();
```
[the remark this answers is not in the document]
Comment (sakthipriyas) [reply psrdada_cpp/effelsberg/edd/src/SKRfiReplacementCuda.cu:47]: get_rfi_window_indices() and get_clean_window_indices() initializes the class members _rfi_window_indices and _clean_window_indices which are device vectors.

File: psrdada_cpp/effelsberg/edd/SpectralKurtosisCuda.cuh [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+public:
+    /**
+     * @brief      constructor
+     *
+     * @param(in)  nchannels     number of channels.
+     * @param(in)  window_size   number of samples per window.
+     * @param(in)  sk_min        minimum value of spectral kurtosis.
+     * @param(in)  sk_max        maximum value of spectral kurtosis.
+     */
+    SpectralKurtosisCuda(std::size_t nchannels, std::size_t window_size, float sk_min = DEFAULT_SK_MIN,
```
[the remark this answers is not in the document]
Comment (sakthipriyas) [reply psrdada_cpp/effelsberg/edd/SpectralKurtosisCuda.cuh:36]: This is not clear to me. Do you mean that the input stream should be passed to the constructor and has to be stored as a class member for further processing? In the implementation, the input stream is passed in compute_sk() function to avoid any device copies.

File: psrdada_cpp/effelsberg/edd/SpectralKurtosisCuda.cuh [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+#include <thrust/iterator/transform_iterator.h>
+#include <thrust/iterator/discard_iterator.h>
+#include <thrust/execution_policy.h>
+#include <nvToolsExt.h>
+
+namespace psrdada_cpp {
+namespace effelsberg {
+namespace edd {
+
+#define DEFAULT_SK_MIN 0.9
```
[the remark this answers is not in the document]
Comment (sakthipriyas) [reply psrdada_cpp/effelsberg/edd/SpectralKurtosisCuda.cuh:17]: I could make them as class default variables. But I am still trying to figure out how to make sk_min and sk_max as optional parameters in constructor. But, is this what you meant?

File: psrdada_cpp/effelsberg/edd/src/SKRfiReplacementCuda.cu [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+    nvtxRangePop();
+    compute_clean_data_statistics();
+}
+
+void SKRfiReplacementCuda::compute_clean_data_statistics() 
+{
+    nvtxRangePushA("compute_clean_data_statistics");
+    std::size_t length = _clean_data.size();
+    _ref_mean = (thrust::reduce(_clean_data.begin(), _clean_data.end(), thrust::complex<float> (0.0f, 0.0f))). real() / length;
+    _ref_sd = std::sqrt(thrust::transform_reduce(_clean_data.begin(), _clean_data.end(), mean_subtraction_square(_ref_mean),
```
[the remark this answers is not in the document]
Comment (sakthipriyas) [reply psrdada_cpp/effelsberg/edd/src/SKRfiReplacementCuda.cu:101]: Both real and imaginary distributions have the same mean and standard deviation. Included this comment in the code.

File: psrdada_cpp/effelsberg/edd/SKRfiReplacement.hpp (hunk 1 above)
Comment (sakthipriyas) [reply psrdada_cpp/effelsberg/edd/SKRfiReplacement.hpp:27]: Yes. This has been fixed in coda version.

File: psrdada_cpp/effelsberg/edd/SKRfiReplacementCuda.cuh (hunk 2 above)
Comment (sakthipriyas) [reply psrdada_cpp/effelsberg/edd/SKRfiReplacementCuda.cuh:18]: Agreed. It doesn't have to be a macro. This reminds me to implement a case when the entire channel is with RFI (100% RFI) and there are no clean windows. In this case, no RFI replacement should happen in the function call.

Comment (sakthipriyas) [reply psrdada_cpp/effelsberg/edd/SKRfiReplacementCuda.cuh:18]: I had that condition (if RFI = 100%) check implemented already.

File: psrdada_cpp/effelsberg/edd/SpectralKurtosisCuda.cuh (hunk 6 above)
[the remark this answers is not in the document]
Comment (TobiasWinchen) [reply psrdada_cpp/effelsberg/edd/SpectralKurtosisCuda.cuh:17]: You should IMO just use     ` SpectralKurtosisCuda(std::size_t nchannels, std::size_t window_size, float sk_min = 0.9, ...)  ` and not have an extra define. I don;t see any benefit,. It makes the code harder to read and bears the risk that it is used in the code somewhere where it shouldn't be used but the value of the sk_min variable should be used.

File: psrdada_cpp/effelsberg/edd/SpectralKurtosisCuda.cuh (hunk 5 above)
[the remark this answers is not in the document]
Comment (TobiasWinchen) [reply psrdada_cpp/effelsberg/edd/SpectralKurtosisCuda.cuh:36]: The actual kernel calls as e.g.: `compute_sk_kernel<<<gridSize, blockSize>>>(...)`  should use a user defined stream `compute_sk_kernel<<<gridSize, blockSize, sharedmemsize, STREAM >>>(...)`

**Review by TobiasWinchen — CHANGES_REQUESTED** [review]
The SKRfiReplacementCuda class needs some restructuring as indicated below because the duplication of information in internal member variables is unnecessary and introduced a bug.
